src/os.rs

```diff
@@ -520,20 +527,59 @@ mod imp {
 #[cfg(target_os = "redox")]
 mod imp {
     use {Error, ErrorKind};
-    use super::ReadRng;
+    use std::fs::File;
+    use std::io::Read;
+    use std::io::ErrorKind::*;
+    use std::sync::{Once, Mutex, ONCE_INIT};
 
     #[derive(Debug)]
-    pub struct OsRng {
-        inner: ReadRng,
-    }
+    pub struct OsRng();
+
+    // TODO: remove outer Option when `Mutex::new(None)` is a constant expression
+    static mut READ_RNG_FILE: Option<Mutex<Option<File>>> = None;
+    static READ_RNG_ONCE: Once = ONCE_INIT;
 
     impl OsRng {
         pub fn new() -> Result<OsRng, Error> {
-            let reader_rng = ReadRng::open("rand:")?;
-            Ok(OsRng { inner: reader_rng })
+            READ_RNG_ONCE.call_once(|| {
+                unsafe { READ_RNG_FILE = Some(Mutex::new(None)) }
+            });
+
+            // We try opening the file outside the `call_once` fn because we cannot
+            // clone the error, thus we must retry on failure.
+
+            let mutex = unsafe { READ_RNG_FILE.as_ref().unwrap() };
+            let mut guard = mutex.lock().unwrap();
+            if (*guard).is_none() {
+                info!("OsRng: opening random device 'rand:'");
+                let file = File::open("rand:").map_err(|err| {
+                    match err.kind() {
+                        Interrupted => Error::new(ErrorKind::Transient, "interrupted"),
+                        WouldBlock => Error::with_cause(ErrorKind::NotReady,
+                                "opening random device would block", err),
+                        _ => Error::with_cause(ErrorKind::Unavailable,
+                                "error while opening random device", err)
+                    }
+                })?;
+                *guard = Some(file);
+            };
+            Ok(OsRng())
         }
-        pub fn try_fill_bytes(&mut self, v: &mut [u8]) -> Result<(), Error> {
-            self.inner.try_fill_bytes(v)
+
+        pub fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+            if dest.len() == 0 { return Ok(()); }
+            trace!("OsRng: reading {} bytes from random device", dest.len());
+
+            // Since we have an instance of Self, we can assume that our memory was
+            // set with a valid object.
+            let mutex = unsafe { READ_RNG_FILE.as_ref().unwrap() };
```

Comment:
Here the old comment actually makes sense — we have `self` on an object whose constructor initialised the mutex.